### ISSUE TITLE
Improve HAProxy reload

### ISSUE THREAD

**Issue by jcmoraisjr** [jcmoraisjr/haproxy-ingress#1]
Useful links:

* http://serverfault.com/a/580630
* http://www.mail-archive.com/haproxy@formilux.org/msg06885.html
* https://engineeringblog.yelp.com/2015/04/true-zero-downtime-haproxy-reloads.html

**Comment by carlanton** [comment]
GitHub has solved this in a pretty interesting way: 
* https://githubengineering.com/glb-part-2-haproxy-zero-downtime-zero-delay-reloads-with-multibinder/
* https://github.com/github/multibinder

It's much easier to implement compared with fiddling with iptables rules etc. I did a hacky PoC implementation here https://github.com/carlanton/haproxy-ingress/commit/1ee8934a285ec7ab07f6fa62e8f356376d05b20b but I haven't tested it yet :)

**Comment by jcmoraisjr** [comment]
@carlanton thank you! No doubt I'll include your implementation on our tests.

**Comment by thekalinga** [comment]
Any chance of this @carlanton change being included in this codebase?

**Comment by jcmoraisjr** [comment]
@thekalinga This feature will be included in the codebase, no doubt. Perhaps carlanton's implementation, perhaps another, it depends on the result of our tests. Planning to release a beta with this feature in a couple of weeks. I'll send a notification in this same issue.

**Comment by thekalinga** [comment]
@jcmoraisjr Thank you for your response.

I have couple of more questions. Not sure whether I should raise an issue/ask about this.

I'll post here in anycase.

1. Is the compatibility with kube-lego is planned/it already works?
2. Is there any plan to integrate the controller directly with kubernetes API to get the service port information directly so that the kube-proxy can be skipped altogether (I'm assuming that will give better performance), like nginx plus does?

**Comment by jcmoraisjr** [comment]
@thekalinga about your questions:

1. I'm planning to include, test and document kube lego usage, but this has a lower priority

2. The controller already read endpoints from Kubernetes API and HAProxy is configured with pod's IP, without proxies. If you are talking about connections from outside into HAProxy, have a look at #6 .

Feel free to file a new issue if you have any doubt.

**Comment by thekalinga** [comment]
@jcmoraisjr Thanks for the clarification.

Kube Lego would increase the appeal of the controller, which means more usage, more hardening, and possibly more contributions.

Thank for your effort in enabling haproxy integration with kubernetes in anycase.

**Comment by thekalinga** [comment]
@jcmoraisjr Does this mean, you also auto detect port number on the pods?

Currently, nginx plus uses DNS SVC records for the service to identify each pod ip & respective port

**Comment by jcmoraisjr** [comment]
@thekalinga Please repost this same question to a new issue, I'll repost my answer there.

---

Taking the Ingress syntax:

```
spec:
  rules:
  - host: your.domain.com
    http:
      paths:
      - backend:
          serviceName: kube-svc-name
          servicePort: 8080
        path: /
```

`serviceName` is used to query all endpoints from kubernetes API, so pod IP address are used. No proxy here.

`servicePort` is the port name or port number to connect to and you should provide it. Ingress will query the `targetPort` from the endpoint and use it. This is an Ingress Core functionality, not a HAProxy Ingress one.

**Comment by thekalinga** [comment]
@jcmoraisjr Sure. Will the whole setup locally & then post any more queries in a new issue. Thanks

**Comment by chlunde** [comment]
Will be fixed upstream in 1.8: https://www.haproxy.com/blog/truly-seamless-reloads-with-haproxy-no-more-hacks/

**Comment by jcmoraisjr** [comment]
Fyi I've just uploaded a starting version of `--reload-strategy` command line param and also a multibinder strategy. Use `--reload-strategy=multibinder` as the very first command line option. The default strategy is `native` so nothing changes. Feel free to update this same issue with your feedback.

There is not an image yet, this feature should be built from the sources. Planning to start beta releases in a few days.

@carlanton thank you very much for your effort.

**Comment by jcmoraisjr** [comment]
@chlunde great, thanks for sharing. =) Nice to know both multibinder and the new HAProxy strategy are very similar.

**Comment by jcmoraisjr** [comment]
Implemented since v0.3-beta.1. The `canary` tag was also updated to this version.
